app/report.go

```diff
@@ -13,33 +13,24 @@ import (
 
 func (a *App) ReportTelemetry(ar *AppRequest) {
 	log.Printf("INF: %s %s Processing", ar.R.Method, ar.R.URL)
-	// retrieve the request body
-	reqBody, err := io.ReadAll(ar.R.Body)
-	if err != nil {
-		ar.ErrorResponse(http.StatusBadRequest, err.Error())
-		return
-	}
-	log.Printf("INF: %s %s reqBody: %s", ar.R.Method, ar.R.URL, reqBody)
 
-	// unmarshal the request body to the request struct
-	var trReq restapi.TelemetryReportRequest
-	err = json.Unmarshal(reqBody, &trReq)
+	trReq, err := getRequestData(ar)
 	if err != nil {
 		ar.ErrorResponse(http.StatusBadRequest, err.Error())
 		return
 	}
 
 	validate := validator.New(validator.WithRequiredStructEnabled())
-	err = validate.Struct(&trReq)
+	err = validate.Struct(trReq)
 	if err != nil {
 		ar.ErrorResponse(http.StatusBadRequest, err.Error())
 		return
 	}
 
-	log.Printf("INF: %s %s trReq: %s", ar.R.Method, ar.R.URL, &trReq)
+	log.Printf("INF: %s %s trReq: %s", ar.R.Method, ar.R.URL, trReq)
 
 	//Save the report into the staging db
-	a.StoreTelemetryReport(reqBody, trReq.TelemetryReport.Header.ReportId)
+	a.StoreTelemetryReport([]byte(trReq.String()), trReq.TelemetryReport.Header.ReportId)
```

Comment:
So in this case you are effectively JSON marshalling the structure again to generate the JSON blob to store in the staging DB.

If we structure your code changes a bit differently we may be able to avoid this and made the solution more generic.

If we split the getRequestData into 2 phases, one to provide the (possibly decompressed) reqBody, and the second to decode it, you can probably implement the first phase as a method on the AppRequest structure...